golang/cosmos/x/swingset/keeper/store_snapshotter.go

```diff
@@ -0,0 +1,601 @@
+// height. If a snapshot is already in progress, this will fail. The snapshot
+// processing is delegated to the provided `snapshotTaker`.
+//
+// The snapshot operation is performed in a goroutine, and the `snapshotTaker`
+// is invoked solely in that goroutine.
+// If the caller requires synchronization, the `snapshotTaker` is guaranteed to
+// be called only between the "started" and "done" events, which can be awaited
+// by calling respectively the `WaitUntilSnapshotStarted` and
+// `WaitUntilSnapshotDone` methods from the thread that initiated the snapshot.
+func (snapshotter *SwingStoreSnapshotter) InitiateSnapshot(height uint64, snapshotTaker SwingStoreSnapshotTaker, exportOptions SwingStoreExportOptions) error {
+	err := snapshotter.checkNotActive()
+	if err != nil {
+		return err
+	}
+
+	encodedExportOptions, err := json.Marshal(exportOptions)
+	if err != nil {
+		return err
+	}
+
+	snapshotHeight := height
+
+	logger := snapshotter.logger.With("height", height)
+
+	// Indicate that a snapshot has been initiated by setting `activeSnapshot`.
+	// This structure is used to synchronize with the goroutine spawned below.
+	// It's nilled-out before exiting (and is the only code that does so).
+	active := &activeSnapshot{
+		height:        snapshotHeight,
+		logger:        logger,
+		startedResult: make(chan error, 1),
+		retrieved:     false,
+		done:          make(chan error, 1),
+	}
+	snapshotter.activeSnapshot = active
+
+	go func() {
+		var err error
+		defer func() {
+			if err != nil {
+				active.done <- err
+			}
+			close(active.done)
+		}()
+
+		action := &snapshotAction{
+			Type:        "COSMOS_SNAPSHOT",
+			BlockHeight: snapshotHeight,
+			Request:     "initiate",
+			Args:        []json.RawMessage{encodedExportOptions},
+		}
+
+		// blockingSend for COSMOS_SNAPSHOT action is safe to call from a goroutine
+		_, err = snapshotter.blockingSend(action, false)
+
+		if err != nil {
+			// First indicate a snapshot is no longer in progress if the call to
+			// `WaitUntilSnapshotStarted` has't happened yet.
+			// Then signal the current snapshot operation if a call to
+			// `WaitUntilSnapshotStarted` was already waiting.
+			active.startedResult <- err
+			close(active.startedResult)
+			logger.Error("failed to initiate swingset snapshot", "err", err)
+			return
+		}
+
+		// Signal that the snapshot operation has started in the goroutine. Calls to
+		// `WaitUntilSnapshotStarted` will no longer block.
+		close(active.startedResult)
+
+		// The user provided `TakeSnapshot` function should call retrieveSnapshot()
+		var retrieveErr error
+		err = snapshotTaker.TakeSnapshot(height, func() error {
+			activeSnapshot := snapshotter.activeSnapshot
+			if activeSnapshot != active || active.retrieved {
+				// shouldn't happen, but return an error if it does
+				return errors.New("snapshot no longer active")
+			}
+
+			retrieveErr = snapshotter.retrieveSnapshot(snapshotTaker.ProcessExportedSnapshot)
```

Comment:
Why does this function directly perform the `blockingSend`s for "initiate" and "discard" but not for "retrieve"?

Reply:
Because of the way state-sync is implemented in the cosmos-sdk. Will try to better details the interleaving.